Replace 'es2015-i18n-tag' w/ 'lingui'

adamsoderstrom opened the pull request:
As the summary says, this PR replaces `es2015-i18n-tag` tooling with `lingui`.

`lingui` comes with a few nice tools and workflows, but also results in SWC support for the Next application in the repo.

**note:** this PR doesn't include migration to SWC. `lingui` supports it, though.

adamsoderstrom committed:
chore(I18n): replace 'es2015-i18n-tag' with 'lingui'

chore(lingui): convert config to 'lingui.config.js'

chore(storybook): apply 'i18n' props to 'RootProvider'

This ensures that every Storybook story has access to the corresponent translations.

build(babel): bring the 'babel.config.js' file back, exclusively for Storybook

build(npm): update 'lingui' devDependencies

The 'yarn extract' script resulted in a memory leak when using a 'lingui.config.js'. Tried to
resolve this with the prior version, but updating solves it.

chore(I18n): add 'ESlint disable' comments

chore(I18n): load and activate translations per Lingui's documentation

adamsoderstrom committed:
chore(Storybook): inline Babel config declaration in Storybook 'main' file

Prior to this commit, we tried to only consume Babel when using Storybook. That was done by moving
the 'babel.config.js' file to the '.storybook' directory. Though, when later updating to the
latest Storybook 7 version, it does no longer work. Hence we strive for using the SWC for Next, we
move the babel config declaration to Storybook's 'main' file.

maeertin commented on src/pages/_app.tsx:
```diff
@@ -38,12 +37,7 @@ function App(props: AppProps) {
-        locale={locale}
-        {...other}
-      >
+      <RootProvider emotionCache={emotionCache} locale={locale} {...other}>
```
So `defaultLocale` is no longer needed? :o

adamsoderstrom replied:
Very good question.
We'll probably need it for monolingual apps?

Will try it out

maeertin replied:
What do you mean? I can't see that `defaultLocale` has been re-added to this file. It needs to be added here for the `defaultLocale` set in `next.config.js` to be forwarded to the I18nProvider.

maeertin reviewed (changes requested):
Sooooo awesome to get this kick-started, great job @adamsoderstrom 🔥✨😄 One thing to consider though is if this PR should contain the whole migration to SWC or if it should only implement lingui. Let's discuss this 👍

adamsoderstrom commented on .eslintignore:
```diff
@@ -1,5 +1,6 @@
 /.git
 /coverage
+/public/locales/*/*.js
```
Ah, that sounds reasonable.

Will try it out!

adamsoderstrom commented on .storybook/main.ts:
```diff
@@ -2,8 +2,54 @@ import { merge } from 'webpack-merge'
 import type { StorybookConfig } from '@storybook/nextjs'
 import webpackBaseConfig from '../webpackBaseConfig'
 
+const productionPlugins = ['@babel/plugin-transform-react-constant-elements']
```
Ah, alright!

Will remove!

adamsoderstrom commented on .storybook/main.ts:
```diff
@@ -2,8 +2,54 @@ import { merge } from 'webpack-merge'
 const config: StorybookConfig = {
   addons: ['@storybook/addon-essentials', '@storybook/addon-a11y'],
+  babel: () => ({
+    presets: [
+      [
+        'next/babel',
+        {
+          // Target ES2015+ browsers
+          'preset-env': {
+            targets: 'Chrome >= 60, Safari >= 10.1, iOS >= 10.3, Firefox >= 54, Edge >= 15',
+            useBuiltIns: false,
+          },
+        },
+      ],
+    ],
```
Ah, true! Will try out removing it

adamsoderstrom replied:
Since i've now reverted the changes in terms of babel, the `preset` array is still present, because Next.js needs them(?)

adamsoderstrom commented on .storybook/preview.js:
```diff
@@ -1,6 +1,7 @@
 import '../scripts/globals'
 import * as React from 'react'
 import { useColorScheme } from '@mui/material'
+import { messages as i18n } from '@lingui/loader!../public/locales/en/messages.po'
```
Ah, yeah. Will try it out

adamsoderstrom commented on src/layouts/AppBase/partials/AppBaseMarketDialog.js:
```diff
@@ -80,15 +77,14 @@ const AppBaseMarketDialog = React.memo(function AppBaseMarketDialog(props) {
-          {t(
-            __translationGroup,
-          )`Please select the country to which your order will be sent. This will give you accurate pricing, delivery times and shipping costs for your destination.`}
+          Please select the country to which your order will be sent. This will give you accurate
+          pricing, delivery times and shipping costs for your destination.
```
Ah, of course! Thanks!

adamsoderstrom committed:
chore(ESLint): add '/public/*' to .eslintignore

adamsoderstrom committed:
chore(Storybook): remove 'productionPlugins'

Hence we rarely do any 'production' builds in regards to Storybook.

adamsoderstrom committed:
chore(babel): move babel config to root of repo

This will result in 'babel' being used and not SWC, for the Next app.

adamsoderstrom committed:
chore(next.config): remove 'swcPlugins'

Hence we are using 'babel' and not SWC.

adamsoderstrom committed:
chore(I18n): add similar import method as used before

Before we used 'Lingui', we used 'es2015-i18n-tag'.
The way we imported the translations were done in c2724a7.
This commit is heavily influenced by that commit,
but instead loading Lingui translations.

adamsoderstrom committed:
chore(pages): remove old 'loadTranslations' function call

adamsoderstrom commented on package.json:
```diff
@@ -29,7 +29,9 @@
     "storybook:test": "test-storybook --url http://127.0.0.1:3001",
-    "typecheck": "tsc --noEmit"
+    "typecheck": "tsc --noEmit",
+    "extract": "lingui extract",
+    "compile": "lingui compile"
```
Ah, yeah. Thinking of `yarn lingui:extract` and `yarn lingui:compile`.

What do you say?

maeertin replied:
Sounds good @adamsoderstrom !

adamsoderstrom replied:
I've now done the above.

adamsoderstrom commented on package.json:
```diff
@@ -60,10 +63,15 @@
+    "@lingui/cli": "^3.17.2",
+    "@lingui/loader": "^3.17.2",
+    "@lingui/macro": "^3.17.2",
+    "@lingui/swc-plugin": "^0.2.3",
```
This should also be removed!

adamsoderstrom commented on public/locales/en/messages.po:
```diff
@@ -0,0 +1,34 @@
+msgid ""
+msgstr ""
+"POT-Creation-Date: 2023-02-07 12:02+0100\n"
+"MIME-Version: 1.0\n"
+"Content-Type: text/plain; charset=utf-8\n"
+"Content-Transfer-Encoding: 8bit\n"
+"X-Generator: @lingui/cli\n"
+"Language: en\n"
+"Project-Id-Version: \n"
+"Report-Msgid-Bugs-To: \n"
+"PO-Revision-Date: \n"
+"Last-Translator: \n"
+"Language-Team: \n"
+"Plural-Forms: \n"
+
+#: src/containers/CartItem/CartItem.js:108
+msgid "Decrease quantity to {0}"
+msgstr ""
+
```
This file should probably also be removed.
Maybe on `npx @noaignite/create-app`?

maeertin replied:
So you mean this file does not need to exist for the app to work?

adamsoderstrom committed:
chore(next.config): add 'webpack' property again

adamsoderstrom commented on next.config.js:
```diff
@@ -12,7 +10,9 @@ const nextConfig = {
     locales: ['en'],
   },
   poweredByHeader: false,
-  webpack: (config) => merge(config, webpackBaseConfig),
```
I've now reverted the removal of the `webpack` property (since we use babel for Next again).

adamsoderstrom committed:
chore(npm): rename 'extract' and 'compile' scripts

adamsoderstrom commented on .storybook/preview.js:
```diff
@@ -62,7 +63,7 @@ function ModeSwitcher(props) {
 export const decorators = [
   (Story, context) => (
-    <RootProvider settings={settings}>
+    <RootProvider settings={settings} i18n={i18n} locale="en">
```
I've now removed the `i18n={i18n}` statement. The import statement as well.

adamsoderstrom commented on src/api/i18n.js:
```diff
@@ -0,0 +1,7 @@
+async function loadTranslations(locale) {
+  const { messages } = await import(`@lingui/loader!../../public/locales/${locale}/messages.po`)
```
Very good question.
I've now changed it to the previous try/catch pattern without actually investigating if it's needed.

adamsoderstrom commented on src/api/i18n.js:
```diff
@@ -0,0 +1,7 @@
+async function loadTranslations(locale) {
```
Ah, totally agree.
I've removed this file now, so we don't need to take any further actions on this.

adamsoderstrom commented on src/api/index.js:
```diff
@@ -1 +1,2 @@
 export * from './utils'
+export { default as loadCatalog } from './i18n'
```
Ah, yeah, totally. My bad. File removed, so this is now resolved.

adamsoderstrom committed:
chore(i18n): remove unwanted translation

adamsoderstrom commented on package.json:
```diff
@@ -60,10 +63,15 @@
     "@babel/core": "^7.14.6",
     "@babel/plugin-transform-react-constant-elements": "^7.10.1",
```
Agree, but since we discussed that we will use Babel for Next.js even after this PR, it will still be necessary.

adamsoderstrom committed:
chore(i18n): add missing translations

adamsoderstrom committed:
chore(npm): remove '@lingui/swc-plugin'

adamsoderstrom commented:
One thing that we still need to do before merging this, is to ensure that we provide a compiled translation file in every production build.

The production builds are currently gitignore'd in this PR (`/public/locales/*/*.js`). I wonder if we shall alter the `yarn build` script to account for this? @maeertin

adamsoderstrom commented:
> One thing that we still need to do before merging this, is to ensure that we provide a compiled translation file in every production build.
> 
> The production builds are currently gitignore'd in this PR (`/public/locales/*/*.js`). I wonder if we shall alter the `yarn build` script to account for this? @maeertin

I think we can skip this, tbh.

[In `next.js`'s own example of using `lingui`](https://github.com/vercel/next.js/tree/canary/examples/with-lingui#with-lingui-example), they state:

> It adds a webpack loader for the messages to avoid having to manually compile while developing as well as adds the compile step to the next build script for production builds.

Webpack loader, in this case, is `@lingui/loader`, i assume.

So we should be fine now! 😊

maeertin commented on src/contexts/Root/index.js:
```diff
@@ -19,14 +19,14 @@ function RootProvider(props) {
-    locale,
     initialSelection,
+    locale,
     settings,
   } = props
 
   return (
     <CacheProvider value={emotionCache}>
-      <I18nProvider defaultLocale={defaultLocale} locale={locale}>
+      <I18nProvider locale={locale} defaultLocale={defaultLocale}>
```
This feels like an unnecessary change? 😯

maeertin commented on src/layouts/App/partials/AppCartDrawer.js:
```diff
@@ -61,7 +61,7 @@ const AppCartDrawer = React.memo(function AppCartDrawer(props) {
       <div>
         <Button component={RouterLink} href="/checkout" variant="contained" fullWidth>
-          {t(__translationGroup)`To checkout`}
+          {t` To checkout`}
```
There's an incorrect whitespace here ;)

maeertin reviewed (changes requested):
Nice further updates, added some more comments :)

maeertin committed:
chore: default the value of locale prop to that of defaultLocale

maeertin committed:
chore(_app.tsx): forward defaultLocale from next to RootProvider

maeertin committed:
chore: add missing translations and fix one typo

maeertin reviewed (approved):
I helped out with some of the comments that I gave from the previous review so I think everything now looks good in this PR 😄 I also noticed Storybook was crashing due a js error regarding i18n so I fixed it.

Is there any more testing that we might want to do before closing and mergin this PR? 🚀 

Awesome job @adamsoderstrom ✨

adamsoderstrom commented:
Ah, hahah! Sweet, @maeertin !
Started working on your comments from today locally, but never got to pushing them! 😄

adamsoderstrom commented:
> Is there any more testing that we might want to do before closing and mergin this PR? 🚀

Yeah. I'll add some comments to the PR and then someone (or me) can have a look on them! 😊
@maeertin

adamsoderstrom commented on lingui.config.js:
```diff
@@ -0,0 +1,15 @@
+/** @type {import('@lingui/conf').LinguiConfig} */
+module.exports = {
+  locales: ['en'],
+  extractBabelOptions: {
+    presets: ['@babel/preset-typescript', '@babel/preset-react'],
+  },
+  sourceLocale: 'en',
```
We should source these from the `next.config.js` file.

```suggestion
  locales: nextConfig.i18n.locales,
  extractBabelOptions: {
    presets: ['@babel/preset-typescript', '@babel/preset-react'],
  },
  sourceLocale: nextConfig.i18n.defaultLocale,
```

maeertin replied:
Totally agree, in a lot of projects we add a `locales.js` file in the root of the project that is then used by `next.config.js` and a potential Sanity setup. This file could then be used for `lingui.config.js` as well :)

adamsoderstrom replied:
Ah, yeah. I've implemented this now! @maeertin

adamsoderstrom commented on src/contexts/I18n/index.js:
```diff
@@ -13,35 +14,25 @@ export function useI18n() {
-  )
-
-  return <I18nContext.Provider value={contextValue}>{children}</I18nContext.Provider>
+  i18n.load(locale, translations)
+  i18n.activate(locale)
```
I'd say that this solution *works*. Maybe we can wrap it in a `useEffect`? 

My biggest question is regarding what impact it would have by wrapping it in a `useEffect` [as in `next.js/examples/with-lingui`](https://github.com/vercel/next.js/blob/2abc824869e4927b820767c1c416a381c8ae528d/examples/with-lingui/pages/_app.js#L15-L24)

maeertin replied:
Let's try it out when we have the chance :)

thekip replied:
Example on nextjs side is broken. Please use examples from lingui repo. With example from nextjs repo you will never get SSR working.

adamsoderstrom reviewed (commented):
Added two comments.
One that i think we should act upon before merging (sourcing locales from `next.config.js` in `lingui.config.js`). The other one is not that important, i'd say.

adamsoderstrom committed:
chore(lingui): source locales from 'next.config.js'

adamsoderstrom committed:
chore(lingui): run 'yarn lingui:extract'

adamsoderstrom committed:
chore(i18n): remove old unnecessary translation file

adamsoderstrom committed:
chore(npm): remove '@babel/plugin-transform-react-constant-elements'

adamsoderstrom committed:
Revert "chore(npm): remove '@babel/plugin-transform-react-constant-elements'"

This reverts commit 5e0750dfb71415c69c296fc7dcb71375ad64f6b6.

adamsoderstrom committed:
chore(i18n): add lingui to '404.js'

alexanderflink reviewed (approved):
This looks great! Thanks @adamsoderstrom 😀

adamsoderstrom committed:
chore(I18n): load plurals

maeertin commented:
You can now run `npx @noaignite/create-app my-app --branch feature/lingui-swc` to test it out in a fresh install :)

In the case that it doesn't work, you might need to delete the npx cache found at `~/.npm/_npx`.

@adamsoderstrom

adamsoderstrom commented:
> You can now run npx @noaignite/create-app my-app --branch feature/lingui-swc to test it out in a fresh install :)
>
> In the case that it doesn't work, you might need to delete the npx cache found at ~/.npm/_npx.

Worked good after removing `~/.npm/_npx`, @maeertin!
According to me, things looks good after building the application as well. 😊

maeertin commented:
Are we ready for merge you think @adamsoderstrom? :)

adamsoderstrom committed:
Merge branch 'develop' into feature/lingui-swc

adamsoderstrom commented:
> Are we ready for merge you think @adamsoderstrom? :)

Yeah, i guess! 😊 @maeertin

maeertin commented:
Actually what's missing in this PR is to entirely clean up the previous solution. Don't know what remains might be in the project from it but we should at least remove the old npm packages `es2015-i18n-tag` and `babel-plugin-i18n-tag-translate`.

adamsoderstrom commented:
> Actually what's missing in this PR is to entirely clean up the previous solution. Don't know what remains might be in the project from it but we should at least remove the old npm packages `es2015-i18n-tag` and `babel-plugin-i18n-tag-translate`.

True. Will remove `es2015-i18n-tag` and `babel-plugin-i18n-tag-translate`.
I've done some cleanup, but will look into more "leftovers"

@maeertin

maeertin commented:
Also just found that `babel-plugin-i18n-tag-translate` is still used in `babel.config.js` :) @adamsoderstrom

maeertin commented:
Let's also remove the old `I18nContext` and bump lingui to v4, which removes the need for `make-plural/plurals` as we discussed @adamsoderstrom

adamsoderstrom committed:
build(deps): remove 'es2015-i18n-tag' and 'babel-plugin-i18n-tag-translate'

Since we've migrated from 'es2015-i18n-tag' plugin to 'lingui', we don't need them as deps anymore.

adamsoderstrom committed:
build(deps): bump lingui from 3.17.0 to 4.0.0

adamsoderstrom commented:
There seems to be a JSX-related bug in the `4.0.0` version of Lingui. Created an issue (https://github.com/lingui/js-lingui/issues/1638)

adamsoderstrom committed:
chore(lingui): add custom extractor (for JSX in .js files)

adamsoderstrom commented:
Seems like [it wasn't a bug](https://github.com/lingui/js-lingui/issues/1638#issuecomment-1537940643)!

I've now made a following commit (), which strives to follow @thekip's instructions.
May be some refactoring to be done (use default + custom extractor, maybe).

@maeertin

thekip commented on src/containers/CartItem/CartItem.js:
```diff
@@ -51,7 +52,6 @@ function CartItem(props) {
   const { product } = cartItem
 
   const { decreaseItem, increaseItem, removeItem } = useCentraHandlers()
-  const { t } = useI18n()
```
You have to use useLingui hook if you are using `t` macro inside a react components. Otherwise component will not be updated when language is switched.  Or use <Trans> react macro, which will automatically subscribed for changes.  

So for smoother migration i will suggest to replace all `const { t } = useI18n()` to `useLingui()` (even if the result of it is not used)

thekip commented:
Guys, I will add my 50 cents here.

1. the current state of your integration will not support Next's  SSR / SSG. I didn't see any `getServerSideProps` / `getStaticProps`. Check examples: https://github.com/lingui/js-lingui/tree/main/examples and working nextjs project here https://github.com/alt3/rank-my-wallet
2. I recommend to use `<Trans>Hello {name}</Trans>` whenever possible. `t` macro is confusing and easy to misuse. 
3. You also can leverage an automatic page level catalog splitting by using "[experimental extractor](https://lingui.dev/guides/message-extraction#dependency-tree-crawling-experimental)", here is an example https://github.com/alt3/rank-my-wallet/pull/148 you still will have issues with JSX inside JS files, because it uses esbuild under the hood, which most likely will not understand it.

adamsoderstrom committed:
Revert "chore(lingui): add custom extractor (for JSX in .js files)"

This reverts commit e021ab609489efe10b6b208c968b31d6a407bd8c.

adamsoderstrom committed:
build(deps): bump lingui to ^4.2.1

Since the new version has built-in support for jsx in ".js" files, we don't need the custom
extractor (which was reverted in the previous commit).

maeertin commented:
Closing PR as "cia" is now purely a documentation app.
